Gbagfx 2n chunk handling

Retnuhytnuob opened the pull request:
## Description
Tooling support for graphics with 2^n based chunking.
 - This handles the tile reordering to optimize into chunks with '2^n stepdown' sizes.
 - This handles both encoding and decoding. (I've confirmed that the output binary matches the original)
 - Includes a breakout of the different pieces of the graphics/pokedex/sprites.png for demonstration purposes, including: (not limited to; mentioned for demonstration purposes)
1. The 13x8 'start transmission' prompt, with chunks of size: 8x8, 4x8, 1x8.
2. The 7x2 'Press Select' section, with chunks of size: 4x2, 2x2, 1x2
3. The 3x3 'azuril' animation, with chunks of size: 2x2, 1x2, 2x1, 1x1.
4. Note: this 3x3 form _matches_ the ordering of the hatch sprites. I've included one as an example. (If we desire to, we could deprecate that 'hatch specific' option/language of the gbagfx tool.)
5. The 6x4 sprite for the 'pokemon not seen' picture. (note: this is done as 2x2 sprites, with a width of 6; works with or without the 2^n support)

Usage: 
`gbagfx {src} {output} -mwidth {x} -mheight {y} -map-2n` with optional `-width {x2}`  on the .4bpp -> .png mode. (x2 being a multiple of x, for a more horizontal layout of the sprite sheet, rather than pure vertical.) and optional `-palette {.gbapal}` for palette data.

I've included a txt document of how the 'breakout' graphics relate to the original sprites.png graphic. - I _have not_ touched the make file to have our code use this yet. We'd need some way of indicating the x/y tile size for the individual graphic, and if it uses this mode.

**Considerations before merging:**
 1. While this _does_ allow encoding/decoding with the strategy we've seen, there are some outliers that don't currently match the behavior... 
 - See, for example, the link_gbasp sprite. Rather than the vertical x2 ... x1 chunking that we see in the 3x3 sprites, (and in the background for the dusclops stage completion 'bonus counter',) this one seems to want this 6x3 as a 4x3 and 2x3. 
 - The dusclops bonus counter was done as 4x2, 4x2, 4x2, 2x2; 4x1, 4x1, 4x1, 2x1. This is _basically_ the same as the 2^n parsing, but with a cap at 4, rather than the 8 seen in the 'graphic' part of that same sprite. - Presumably, this was done for the 'thousands separator' chunking.
 - If we can determine a rule for when to use that, rather than the vertical chunk, we can add it into the logic building the 2nTileMapping. 
 - This could be turned into a toggle flag. 
 2. While this _does_ imitate some of what we see in the existing data... (and could be useful for clarity) this _doesn't_ help in getting it _into_ that form. The breakout example was done manually... 
 - Long term, for maintenance/modding this is likely useful, with the extra effort put in. Especially if we want to modify the sprites, while keeping existing sizes. However...
 - There may be a separate 'data mapping', or additional graphics rules that we still to uncover in code, for breaking the 'sprite collection' into its smaller usable chunks. If so, this capability might be obsoleted by that logic.
 3. We may want some cleanup/optimization/renaming of the 2^n chunk mapping support in the gbagfx tool.
 - Currently, I'm initializing the array for the '2^n chunks' in the start of the ConvertFromTiles4Bpp code. (outside the array) There may be a way to make that only used if it's using that mode.
 - There may be a more efficient way to do the 'metatile advance' for this case. (I've prioritized getting it working, rather than optimizing at this stage. - Haven't really coded in C in almost 2 decades.)
 - I've currently left my debug statements for displaying the chunk sizes/indexed mapping, in comments in the file. -- We may want to give a debug toggle to display them, or clean those up from the code, once we are sufficiently confident in it.

## **Discord username**
retnuhytnuob

Cyphgirl commented on graphics/stage/gUnk_0851DB8C_a.png:
I saw this image when I hit the bumpers enough times to trigger clearing all tiles on the catch 'em mode in one shot in the sapphire board.

Retnuhytnuob commented:
Updated the gbagfx oam chunk PR with the catch_mon graphics. If that other memory block Cyphgirl mentioned does contain the palette information, (or if that's an address to where that is stored) I was unable to get it to work correctly. It's possible there's something about my extraction technique that isn't getting it to parse the bytes correctly... Someone else can check that.

Removed the special 6x4 oam case, replacing the 2 json entries that used it with the "2x2, width 6" mode, without the oam flag. This is net the same processing-wise, but means the 'wailmer' I stumbled upon will slice correctly when parsing the 6x4 tile set with the oam flag.

This means there's only 2 remaining 'override' cases for the oam: the 13x8, for 'transmission start' and the 6x3 for the 'gbasp link' - both sub images of pokedex/sprite. Those _don't_ have a corresponding ability to slice correctly without this handlling. If we find other sprites with those dimensions (unlikely?) that need the standard oam slicing handler, we may need to change the override from a default to a pattern that can be fed in.

huderlem commented on graphics_rules.mk:
```diff
@@ -1,6 +1,78 @@
+			if [ -n "$$mw" ] && [ "$$mw" != 0 ]; then mw_arg="-mwidth $$mw"; fi; \
+			if [ -n "$$mh" ] && [ "$$mh" != 0 ]; then mh_arg="-mheight $$mh"; fi; \
+			if [ "$$oam" = true ]; then oam_arg="-oam"; fi; \
+			echo -e "Processing segment $$seg: $(GFX) $*_$$seg.png $*_$$seg.4bpp $$mw_arg $$mh_arg $$oam_arg; \n"; \
```
The several `echo` commands seem to add significant extra time when I run `make`.  Can we comment them out?  i imagine they're useful for debugging.

nathan-schutz-23310 replied:
Feel free to comment/clean them out. They were absolutely something I needed to confirm it was passing the flags correctly, but may only be needed for debugging at this point.

huderlem commented on graphics_rules.mk:
```diff
@@ -10,8 +82,8 @@ graphics/options/sprites.4bpp: graphics/options/sprites.png
 
-$(MON_HATCH_SPRITES_DIR)/%.4bpp: $(MON_HATCH_SPRITES_DIR)/%.png
-	$(GFX) $< $@ -pinball-hatch-sprite
+#$(MON_HATCH_SPRITES_DIR)/%.4bpp: $(MON_HATCH_SPRITES_DIR)/%.png
```
We can obliterate the old hatch-sprite code now.

huderlem commented on graphics_rules.mk:
```diff
@@ -1,6 +1,78 @@
+#	oam=$$(echo $$params | grep -q 'oam=true' && echo "-oam"); \
+#	$(GFX) $< $@ -mwidth $$mw -mheight $$mh $$oam
+
+
```
There is one big issue with the Makefile rules.  Modifying an individual metasprite's PNG file does not cause it to be reprocessed the next time you run `make`.

For example:
1. Compile ROM
2. Make some changes to `graphics/pokedex/sprites_select_left.png`.
3. Compile ROM again
4. Observe that the changes to the PNG are not present in the ROM.

To fix this, we need to specify Makefile dependency rules between the target PNG and oam segment PNGs.  If we were to manually specify them in the Makefile, it would look like this:
```
graphics/pokedex/sprites.png: graphics/pokedex/sprites_select_left.png graphics/pokedex/sprites_medal_gem.png etc.
graphics/titlescreen/sprites.png: graphics/titlescreen/sprites_saved_game.png graphics/titlescreen/sprites_no_saved_game.png
etc....
```
Of course, we'd actually want to generate those automatically by reading the JSON files and outputting those dependencies.  You can use Makefile's [eval](https://www.gnu.org/software/make/manual/html_node/Eval-Function.html) function to define those dependency rules.  It might be easiest to write a little separate script that processes the JSON and loops, and `eval` would invokeit.

Retnuhytnuob replied:
I believe I might have the segment update logic working. Updating one of the segments should now trigger the corresponding .4bpp file to update.

However... one thing I saw while testing... If the file timestamp is _newer_, it's able to rerun the build, picking up that change. But if you then try to replace that change with an older copy of the file, with an older timestamp, it doesn't see that as a change that triggers needing to recompile the .4bpp (and/or the whole rom inc chain) ... ChatGPT says this is normal for the make process.

I _was_ able to use powershell to update the timestamp on the reverted file, and when I made that change, it did pick it up as a change, and recompiled on the basic make, with a successful compare. 

If this is "as expected", everything should be good to go. If not, it may need some tweaking.

huderlem reviewed (commented):
Overall, this generally looks good for now.  I think we need to address the sprite-segment Makefile dependency issue before merging, though.  There are definitely some gaps and improvements that we can make later, of course.

nathan-schutz-23310 commented on graphics/mon_hatch_sprites/aron_hatch2.png:
Correct. Can be removed

nathan-schutz-23310 commented on graphics/mon_catch_sprites/catch_sprites_gfx.json:
```diff
@@ -0,0 +1,350 @@
+  "files": [
+    {
+      "f": "catch_mon_01_treecko",
+      "tileCount": 108
```
No. It's not _used_ by any of the tooling. 

The reason it's there is that I needed to narrow down the possible X*Y space to try, so I had the init tool give a count of tiles when it was generating the json from the list of 4bpp in the folder. Once the size has been determined, we could clean it out of the Json data entirely. 

Given what it's there for, and how it gets set, it doesn't make sense to have as a default, unless we want that for reference later for some reason.

Retnuhytnuob replied:
No. It's not used by any of the tooling.

The reason it's there is that I needed to narrow down the possible X*Y space to try when sizing any given image, so I had the init tool give a count of tiles when it was generating the json from the list of 4bpp in the folder. Once the size has been determined, we could clean it out of the Json data entirely.

Given what it's there for, and how it gets set, it doesn't make sense to have as a default, unless we want that for reference later for some reason.

Retnuhytnuob commented on graphics_rules.mk:
```diff
@@ -1,6 +1,78 @@
+
+
+# 2. Pattern rule for .4bpp files using params.txt
+# Convert *.png -> *.4bpp using optional parameters from graphics/params.txt
```
I renamed it. Params.txt is too generic, might conflict with other uses. I need to clean up the comments.

Retnuhytnuob commented on graphics/stage/ruby/sharpedo.png:
If we have tools that handle the sharpedo, I'm not sure what it needs to be fed on the gbagfx call.

This isn't a rectangle overall. There's a 'top fin' 1x2 and a single to the right as a 1x1 that are separate from the main body.

While I could slice this into 9 separate pieces (top fin, body, side fin) for the 3... I'd almost prefer leaving it in its 'straight from the memory' format until we have proper tooling support for that 'oddly shaped' object.


----
As for how to potentially deal with it... it might be possible to set a 'override configuration' as a flag it sends separate from the oam/rectangle size. Internally, I think we could use the 'oam chunks' to process though the pieces, just with a custom pattern defined somewhere. If we want to try that approach, I'd probably make the 2 'override' instances we found use that flag, rather than a direct 'size determined oam' override.
